core/templates/dev/head/pages/exploration_player/AudioPreloaderServiceSpec.js

```diff
@@ -0,0 +1,279 @@
+                  }
+                ],
+                correct: false,
+                outcome: {
+                  feedback: [
+                    "<p>Let's go to State 1</p>"
+                  ],
+                  dest: 'State 1',
+                  param_changes: []
+                }
+              },
+              {
+                rule_specs: [
+                  {
+                    inputs: {
+                      x: '2'
+                    },
+                    rule_type: 'Contains'
+                  }
+                ],
+                correct: false,
+                outcome: {
+                  feedback: [
+                    "<p>Let's go to State 2</p>"
+                  ],
+                  dest: 'State 2',
+                  param_changes: []
+                }
+              }
+            ],
+            hints: []
+          },
+          classifier_model_id: null
+        }
+      },
+      param_specs: {},
+      param_changes: [],
+      version: 1
+    };
+
+    requestUrl1 = UrlInterpolationService.interpolateUrl(
+      '/audiohandler/<exploration_id>/audio/<filename>', {
+        exploration_id: '1',
+        filename: 'en-0.mp3'
+      });
+    requestUrl2 = UrlInterpolationService.interpolateUrl(
+      '/audiohandler/<exploration_id>/audio/<filename>', {
+        exploration_id: '1',
+        filename: 'en-1.mp3'
+      });
+    requestUrl3 = UrlInterpolationService.interpolateUrl(
+      '/audiohandler/<exploration_id>/audio/<filename>', {
+        exploration_id: '1',
+        filename: 'en-2.mp3'
+      });
+    requestUrl4 = UrlInterpolationService.interpolateUrl(
+      '/audiohandler/<exploration_id>/audio/<filename>', {
+        exploration_id: '1',
+        filename: 'en-3.mp3'
+      });
+  }));
+
+  it('should maintain the correct number of download requests in queue',
+    function() {
+      $httpBackend.expect('GET', requestUrl1).respond(201, 'audio data 1');
+      $httpBackend.expect('GET', requestUrl2).respond(201, 'audio data 2');
+      $httpBackend.expect('GET', requestUrl3).respond(201, 'audio data 3');
+
+      var exploration = eof.createFromBackendDict(explorationDict);
+      aps.init(exploration);
+      atms.init('en', 'en', 'en');
+      ecs.setExplorationId('1');
+      aps.kickOffAudioPreloader(exploration.getInitialState().name);
+
+      expect(aps.getFilenamesOfAudioCurrentlyDownloading().length).toBe(3);
+      expect(aps.isLoadingAudioFile('en-0.mp3')).toBe(true);
+      expect(aps.isLoadingAudioFile('en-1.mp3')).toBe(true);
+      expect(aps.isLoadingAudioFile('en-2.mp3')).toBe(true);
+      expect(aps.isLoadingAudioFile('en-3.mp3')).toBe(false);
+      $httpBackend.flush(1);
+      expect(aps.getFilenamesOfAudioCurrentlyDownloading().length).toBe(3);
+      $httpBackend.flush(1);
+      expect(aps.getFilenamesOfAudioCurrentlyDownloading().length).toBe(2);
+      $httpBackend.flush(1);
+      expect(aps.getFilenamesOfAudioCurrentlyDownloading().length).toBe(1);
```

Comment:
I had some trouble here. I want to do one more $httpBackend.flush(1) because there should be one more pending download, but for some reason it says there aren't any requests to flush if I try to do that.

Reply:
I'm investigating this now, I'll get back to you!

Reply:
In the error log there's a UrlInterpolationService is not defined (if you scroll higher up). That might be the root cause. Still investigating.

Reply:
OK, it's unrelated. But I figured it out -- it's because you didn't declare a GET for requestUrl4 in line 240 above.

Also, you could do the final flush after line 261 if you want to verify that the intermediate results are correct.

Reply:
That's weird... I thought that was the first thing I tried but was getting an 'Unsatisfied requests' error, I guess I must've changed something in the meantime that fixed it. Thanks!

Reply:
Done